types/node/index.d.ts

```diff
@@ -4379,6 +4460,22 @@ declare module "fs" {
      */
     export function readFileSync(path: PathLike | number, options?: { encoding?: string | null; flag?: string; } | string | null): string | Buffer;
 
+    export type WriteFileOptions = { encoding?: string | null; mode?: number | string; flag?: string; } | string | null;
+
+    /**
+     * Asynchronously writes data to a file, replacing the file if it already exists.
+     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
+     * URL support is _experimental_.
+     * If a file descriptor is provided, the underlying file will _not_ be closed automatically.
+     * @param data The data to write.
+     * @param options Either the encoding for the file, or an object optionally specifying the encoding, file mode, and flag.
+     * If `encoding` is not supplied, the default of `'utf8'` is used.
+     * If `mode` is not supplied, the default of `0o666` is used.
+     * If `mode` is a string, it is parsed as an octal integer.
+     * If `flag` is not supplied, the default of `'w'` is used.
+     */
+    export function writeFile(path: PathLike | number, data: BinaryData, options: WriteFileOptions, callback: (err: NodeJS.ErrnoException) => void): void;
```

Comment:
An overload using `BinaryData` doesn't add any value as the other overload has `any`. `any` instead of `string` is ok here as node actually converts the data via `Buffer.from('' + data, options.encoding || 'utf8')`

Reply:
Ideally we would create an interface that has `.toString()` in it, the idea was to signal that these types are _preferred_.

Reply:
Even `undefined` and `null` works here which have no `.toString()`. Risk is high that this breaks existing code.

Reply:
I guess, although the chance of it breaking _anything_ is extremely low as it is an overload and TS will probably choose the other option if the `BinaryData` one doesn't match.

I don't feel strongly on this at all but would like to have another opinion on this.

Reply:
Problem with typing the _preferred_ or _intended_ type is that users have to cast if they use something else which is not intended but works.

Reply:
Also, to be fair, I get why it was originally typed as `any`, but I we should try to differentiate between `intended` and `allowed` usage here.

Passing `BinaryData`, `string` etc. is the intended behaviour, maybe even `number` but at the point where you hit `null` or `undefined` it's extremely likely to unintended behavior and should probably be prevented, that's the entire point of TS after all.

Again, I don't feel strongly about this but I think providing an overload with the _intended_ types is at least an indicator of incorrect usage.

Reply:
To be clear, I don't want to force any significant changes and I am likely to remove this as it's not really aligned with the rest of the node typings, but having a brief conversation would be nice :)

Reply:
In general I agree here but I have seen several times that people complain about breaking changes similar to this.
I think adding the intended type in the doc part (like in node docs) is better as this is what you see during typing.

Anyhow,  I will not block both variants; maybe other contributors give their vote.